### ISSUE TITLE
"Error connecting to PLC: Error: read ECONNRESET"

### ISSUE THREAD

**Issue by Spawn2qc** [netsmarttech/node-red-contrib-s7#18]
I have communicate between S7-1200/S7-1500 and Node-RED on IOT2040 many times in the past.

Yesterday, I installed a brand new image V2.1.3 on a IOT2040, change IP address to my needs, auto-start Node-red, installed S7 node and finally reboot before doing my flow.

I usually use the "Rack/slot" mode (not TSAP) for the "s7 endpoint node" and access inputs/outputs and variables in DB without any issue.

Here is the current configuration to communicate with a S7-1200:

![1](https://user-images.githubusercontent.com/33728856/36805326-1a1b0fb6-1c8b-11e8-9edc-a8d2138fdaad.PNG)
![2](https://user-images.githubusercontent.com/33728856/36805336-1f4d0192-1c8b-11e8-8ef6-bda06159918b.PNG)

Here is the DB:
![3](https://user-images.githubusercontent.com/33728856/36805342-25f64c88-1c8b-11e8-9a8a-e365550d6c78.PNG)

Yes, I check the checkbox "Permit access with PUT/GET communication from remote partner".
![4](https://user-images.githubusercontent.com/33728856/36805353-3002c530-1c8b-11e8-8e8b-e3f9330d423e.PNG)

I can see my data is receive by the S7 node, but my debug window is fill with the error "Error connecting to PLC: Error: read ECONNRESET".
![5](https://user-images.githubusercontent.com/33728856/36805370-3c29cce6-1c8b-11e8-8a12-df3049409173.PNG)

I never had this error in the past.

I receive my data, that's not the end of the world, but it is annoying to have a filled debug window.

I tried to add a S7 connection with an "Unknown partner" with IP address of the IOT2040 and use the TSAP mode, but get the same result. 

Do you have an idea what's wrong?

It must be related to the latest version of S7 node 1.3/1.4 because I never got this error in the past and I am doing the same thing.

**Comment by gfcittolin** [comment]
Hi @Spawn2qc 

I think the issue might be with a connection leak we're having under some circumstances. In the case we have a connection problem, we currently ask the library to connect again, but without creating a new library instance, and then the connection leaks

It might be in your case that the PLC is not accepting connections anymore because of these leaked connection, and therefore it logs these `ECONNRESET` errors.

I've been testing the fix for this in the past weeks, and it looks like the problem is solved. I'll publish a new version either today or tomorrow, so you can update your node and check if the problem has been fixed

**Comment by gfcittolin** [comment]
Just published 1.5.0 that might fix the issue. Could you please update and check it?

**Comment by Spawn2qc** [comment]
Hi @gfcittolin 
I think you are right about the PLC not accepting connections anymore.
This morning, I reboot all the system and the S7-1200 must have reset its connection because there is no more error in the debug window.
Yesterday, I reboot Node-RED only, not the S7-1200.
So, maybe, the "leaked" connection in the S7-1200 were still count.
There is 8 S7 connections limit, maybe it was bust.
![1](https://user-images.githubusercontent.com/33728856/36927831-5437e498-1e4e-11e8-82b8-45ae7c713678.PNG)
I will update S7 node to 1.5.0 monday morning and let you know if something goes wrong, else it is solve.
Thank you very much for your support, it is very appreciate!

**Comment by gfcittolin** [comment]
My pleasure. Let me know if it works now so we can close this issue.

**Comment by zoujagger** [comment]
Hi every body,
I am having exactly the same issue using the latest release of the node-red-contrib-s7 1.5.0. 
Indeed, in one hand i am getting the error : "Error connecting to PLC: error: read ECONNRESET" continuously and in the other hand, i receive the data well! However, after a certain period of time, the node-red crashes.  

Please find below the main configurations my architecture.

configuration:
PLC S7-1200
node red 0.18.4
node-red-contrib-s7 1.5.0
OS Ubuntu 16.0.4

Does anybody  have an idea or can help!? It's an emergency ... Thanks

Best regards

**Comment by gfcittolin** [comment]
That is bad news. Could you please:
 - Check the connection resources of your S7-1200 and get the available connections, to confirm our hypothesis that we're leaking connections
 - Post here the stack trace of Node-RED when it crashes, so we can track down where it fails
 - Enable the debug mode on the PLC config on node-red, and then try to simulate the issue. Then get the logs of node-red and post here too, so we can analyze what's wrong.

**Comment by gfcittolin** [comment]
Just released 1.5.1, there was another connection leak case I could simulate here. The leak happened when the node was trying to reconnect to the PLC, and the flow was redeployed. Could you guys please test it out?

**Comment by zoujagger** [comment]
Thanks for your contribution, We are test it and we'll give you the result

**Comment by zoujagger** [comment]
I tested the 1.5.1 release  yesterday, and everything works fine. I no longer have the error message. 
Thanks  for  your reactivity. 

Best regards

**Comment by gfcittolin** [comment]
Thanks for testing!

**Comment by lukaku75** [comment]
Hi,
I have descripted problem above!
But I can't get data from my plc.
The node worked fine for one week but then I get these Errors:
![grafik](https://user-images.githubusercontent.com/50985785/58317915-10a35d00-7e17-11e9-8087-065b25293875.png)
I do several things:
-Reset the SPS
-uninstall the node and install it new
-upgrading node-red.
- I  checked the Ressources
![grafik](https://user-images.githubusercontent.com/50985785/58318198-96270d00-7e17-11e9-8bf2-69494277baa1.png)
-I checked the mechanism
![grafik](https://user-images.githubusercontent.com/50985785/58318275-ba82e980-7e17-11e9-827b-316aa4bb052c.png)

but the ERROR is still there.
What else can I do?
Thanks!
